### ISSUE TITLE
Tracking Issue for "C-unwind ABI", RFC 2945

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#74990]
This is a tracking issue for the RFC "C-unwind ABI" (rust-lang/rfcs#2945).

The feature gate for the issue is `#![feature(c_unwind)]`.

This RFC was created as part of the ffi-unwind project group tracked at https://github.com/rust-lang/lang-team/issues/19.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement the RFC
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Implementation notes

Major provisions in the RFC:

* [x] Add a `C-unwind` ABI and `system-unwind` (I think), we may need more variants.
* [ ] For external functions with the C ABI, we already (I believe) add the "nounwind" LLVM attributes when we build them. We want to continue doing this.
* [ ] For external functions with the C-unwind ABI, we do *not* want any such attributes.
* [ ] For Rust functions defined with the C ABI, e.g., `extern "C" fn foo() { ... }`, we want to force them to abort if there is a panic. The MIR helper function [`should_abort_on_panic`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir_build/build/fn.should_abort_on_panic.html) already exists and I think will modify the generated code to insert the required shims in such a case. They should also be marked as "nounwind" in LLVM, if they're not already. 
* [ ] Rust functions with the "C-unwind" ABI should **not** abort on panic.
* [ ] Use ABI to guide the "nounwind" attribute on callsites as well.
* [ ] Write suitable codegen tests to check generated LLVM IR.

### Unresolved Questions

None. The unresolved questions in the RFC were meant to be solved by future RFCs building on this one.

### Implementation history
* Initial implementation: https://github.com/rust-lang/rust/pull/76570
* Additional work (removing `#[unwind]`, adjusting panic handling): https://github.com/rust-lang/rust/pull/86155

**Comment by katie-martin-fastly** [comment]
Hello! I would be interested in helping implement this RFC. :raising_hand: 

(cc: @aturon)

**Comment by nikomatsakis** [comment]
Hey @katie-martin-fastly, that's great news! I'm going to assign you to the issue for now:

@rustbot assign @katie-martin-fastly 

One thing to mention is that most Rust compiler development discussion takes place on the [rust-lang zulip](https://rust-lang.zulipchat.com/), and this project in particular is chatting in `#project-ffi-unwind`. I or @Amanieu are probably reasonable people to ping with questions. Also, for general "getting started" tips on building and testing the Rust compiler, the [rustc-dev-guide](https://rustc-dev-guide.rust-lang.org/) is your friend. Also, rust-analyzer works well for IDE support (e.g., with vscode), though it requires a small bit of configuration since rustc doesn't build with cargo but rather the `x.py` script. (I couldn't find any documentation on that, I'll ping a few others and see...)

That said, in terms of mentoring and getting started, I think the idea is going to be to adapt an existing mechanism that the compiler offers. We have this attribute `#[unwind(allowed)]` which allows users to mark functions as permitting unwinding -- that mechanism is basically going to be deprecated and replaced with this new ABI, though I think it'd be best to add the ABI before trying to adapt the `unwind` attribute code. Still, searching for code related to that attribute will give you a good idea of what needs to change. 

I can do more mentoring instructions later but, as a starting point, I updated the head post with some of the major goals of the RFC (see the "Implementation notes") section. Also here are some tips to get you started in terms of reading into the rustc code:

* The [`Abi`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_llvm/abi/enum.Abi.html) enum lists out all the known ABIs (there may be more similar enums, not sure). We'll want to extend it with `C-unwind` and perhaps other unwind variants. One way to do this might be to change the `C` variant to `C { unwind: bool }` or something like that, or maybe just add a `Cunwind`, not sure.
* The [`UnwindAttr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_attr/enum.UnwindAttr.html) type records what kind of unwind attribute is placed on a particular function (is unwinding [allowed](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_attr/enum.UnwindAttr.html#variant.Aborts)? does it [force an abort](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_attr/enum.UnwindAttr.html#variant.Aborts)?)
* The [`find_unwind_attr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_attr/fn.find_unwind_attr.html) function checks for what unwind attribute is placed on a particular function.
* The [`should_abort_on_panic`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir_build/build/fn.should_abort_on_panic.html) function is used to decide when a function ought to, well, abort if a panic occurs (as the name suggests).
* The [`can_unwind`] field of this struct seems to be used to [set the LLVM attribute "nounwind"](https://github.com/rust-lang/rust/blob/cfdf9d335501cc0a53ae69c940095cca7d4be0f8/src/librustc_codegen_llvm/abi.rs#L399-L402). I'm not quite sure where it gets set, but I have to stop now because it's Saturday :P so I'll leave that as an exercise to the reader.

**Comment by BatmanAoD** [comment]
Blocked on https://github.com/rust-lang/rust/issues/97659.

**Comment by madsmtm** [comment]
Should probably also be blocked on https://github.com/rust-lang/rust/pull/92964 (`"C-unwind"` fn-pointers don't implement common traits).

**Comment by lopopolo** [comment]
> Should probably also be blocked on #92964 (`"C-unwind"` fn-pointers don't implement common traits).

@madsmtm @BatmanAoD I've put up a PR to resolve this stabilization blocker:

- https://github.com/rust-lang/rust/pull/101263

**Comment by lopopolo** [comment]
@madsmtm @BatmanAoD: https://github.com/rust-lang/rust/pull/101263 has been merged 🎉 What remains before the `C-unwind` ABI can be stabilized?

Once the next nightly drops I should be able to push up a build of Artichoke Ruby which oxidizes a `setjmp`/`longjmp` unwinding mechanism with `catch_unwind`/`panic` if that would help with a motivating usecase for the stabilization report.

**Comment by BatmanAoD** [comment]
@lopopolo We're ready for stabilization! Discussed here: https://rust-lang.zulipchat.com/#narrow/stream/210922-project-ffi-unwind/topic/weekly.20meeting/near/307806910

**Comment by thomcc** [comment]
Hey, I'm concerned about the fact that the current feature (and presumably the proposed stabilization) controls both the existence of the `"*-unwind"` ABI and whether or not the other ABIs are forbidden from unwinding or not (see https://github.com/rust-lang/rust/blob/15d7556de9babe4c1d8f367de1b827494782bd92/compiler/rustc_middle/src/ty/layout.rs#L1048-L1054).

If stabilized in this way, anybody using this functionality will likely need to implement compiler version (or feature) detection, in order to conditionally switch between `extern "C"` and `extern "C-unwind"` based on availability. They also have to release that in time to be ready day 1 when this hits stable. This isn't exactly hard, but seems pretty rough, especially because that may require breaking API changes (because of changing a function's ABI). Perhaps more importantly, doing this requires knowing about it in advance. Many Rust devs only find out about new features when they hit release channel.

For what it's worth, I work on on [a codebase](https://github.com/tcdi/pgx) that uses `extern "C"` as if it's `extern "C-unwind"`. This stabilization will *probably* cause no problems for us, since we've been prepared for it and [very recently](https://github.com/tcdi/pgx/pull/834)[^1] fixed the last of the design issues around it. Speaking way too optimistically, I don't anticipate many problems with changing this for PGX, but do note: This was not trivial to do, and was greatly helped by having multiple Rust project members who have been paying attention to this issue. I believe most projects using would be in a worse place, and might not find out about this in advance.

So, if we're going to make switching to a feature mandatory in the release that stabilizes it, then *at least* we need to publish a blog advising about this in advance. And even then, I think that it's pretty dodgy for us to do.

**Comment by BatmanAoD** [comment]
As I mentioned in Zulip [in Zulip](https://rust-lang.zulipchat.com/#narrow/stream/210922-project-ffi-unwind/topic/weekly.20meeting), I'm okay with breaking this into two features and stabilizing them separately, but I don't know who would need to sign off on that decision.

**Comment by BatmanAoD** [comment]
@nikomatsakis @joshtriplett Do either of you know who should make the decision on whether to stabilize `"C-unwind"` separately from the new `"C"` behavior? In the interim, `"C"` would behave identically to `"C-unwind"`.

**Comment by RalfJung** [comment]
Sounds like a lang team decision to me. I think this can probably be decided with the FCP for stabilizing the feature. So I'd propose suggest you propose one of these two options, ask for T-lang FCP on that, but add a note in the nomination message that there is an alternative here and we'd like to know which way T-lang prefers.

Do we usually do stabilization FCPs on the tracking issue or the stabilization PR?

**Comment by pnkfelix** [comment]
@rustbot label: I-lang-nominated

see question above

**Comment by RalfJung** [comment]
To give some context to the decision here: currently the c_unwind feature flag has two effects:
- it lets you use `extern "C-unwind"`
- it makes rustc assume that an `extern "C"` is `nounwind`, and generate code accordingly (on stable, we currently treat `extern "C"`  as allowing unwinding)

The question is whether when we unlock `extern "C-unwind"` on stable, we should immediately also do the 2nd part on stable, or whether we should have a transition period where for now, we keep treating `extern "C"`  as unwinding, and wait a few releases before making it `nounwind`.

**Comment by jyn514** [comment]
I'm very out of the loop so there's probably a simple answer, but why not keep the current `extern "C"` (and guarantee it's not UB) and add a new `extern "C-nounwind"` that people can opt-in to? That wouldn't be a breaking change.

**Comment by bjorn3** [comment]
For unwinding into Rust that would be an option. For unwinding out of Rust, that would remain a massive footgun as the other side likely doesn't allow unwinding and thus you get UB anyway. Having `extern "C"` abort when unwinding out of Rust is required to fix this UB.

**Comment by BatmanAoD** [comment]
@jyn514 That was discussed as a possibility some time ago, but was ultimately rejected. Unfortunately I don't have a readily available link to the rationale for choosing to add `C-unwind` rather than adding `C-nounwind`, and I don't really remember what the pros and cons were.

**Comment by nikomatsakis** [comment]
To my mind, the major pros/cons are:

* unwind is only expected and desired in a very niche set of cases
* it is not free in that the generated code for C-unwind is less efficient than C to account for possibility of unwinding
* doesn't solve the case of people unwinding *out* from Rust, as @bjorn3 noted

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting. We *would* like to see the stabilization split, so that we can stabilize `"C-unwind"` now, and switch over `"C"` to mean nounwind in a few versions. The stabilization of `"C-unwind"` should also have the `relnotes` tag and include a note that a future version will switch the behavior of `"C"`.

**Comment by BatmanAoD** [comment]
# Request for Stabilization

This is a request for *partial* stabilization of the `c_unwind` feature,
[RFC-2945][rfc-text]. The newly-introduced ABIs will be stabilized, but the
changes in behavior to existing ABIs (such as `extern "C"`) will remain behind
the `c_unwind` feature gate.

## Summary

This stabilization enables ABI strings ending in `-unwind`; specifically:

- `"C-unwind"`
- `"cdecl-unwind"`
- `"stdcall-unwind"`
- `"fastcall-unwind"`
- `"vectorcall-unwind"`
- `"thiscall-unwind"`
- `"aapcs-unwind"`
- `"win64-unwind"`
- `"sysv64-unwind"`
- `"system-unwind"`

The behavior for unforced unwinding (the typical case) is specified in [this
table from the RFC][rfc-table], and has not been changed since the RFC was
accepted. To summarize:

Each ABI is mostly equivalent to the same ABI without `-unwind`, except that
the behavior is defined to be safe when an unwinding operation (`panic` or C++
style exception) crosses the ABI boundary. For `panic=unwind`, this is a valid
way to let exceptions from one language unwind the stack in another language
without terminating the process (as long as the exception is caught in the same
language from which it originated); for `panic=abort`, this will typically
abort the process immediately.

For initial stabilization, *no change* will be made to the existing ABIs. The
existing ABIs have soundness holes, though, which will be fixed when the
behavior described in the RFC is stabilized. Thus, we would like to stabilize
the remainder of the `c_unwind` feature soon. We are splitting the
stabilization into separate releases to give users time to adopt the new ABIs.

## Examples

### Rust `panic` with `"C-unwind"`

```rust
#[no_mangle]
extern "C-unwind" fn example() {
    panic!("Uh oh");
}
```

This function is now permitted to unwind C++ stack frames.

```
[Rust function with `catch_unwind`, which stops the unwinding]
      |
     ...
      |
[C++ frames]
      |                           ^
      | (calls)                   | (unwinding
      v                           |  goes this
[Rust function `example`]         |  way)
      |                           |
      +--- rust function panics --+
```

If the C++ frames have objects, their destructors will be called.

### C++ `throw` with `"C-unwind"`

```rust
#[link(...)]
extern "C-unwind" {
    // A C++ function that may throw an exception
    fn may_throw();
}

#[no_mangle]
extern "C-unwind" fn rust_passthrough() {
    let b = Box::new(5);
    unsafe { may_throw(); }
    println!("{:?}", &b);
}
```

A C++ function with a `try` block may invoke `rust_passthrough` and `catch` an
exception thrown by `may_throw`.

```
[C++ function with `try` block that invokes `rust_passthrough`]
      |
     ...
      |
[Rust function `rust_passthrough`]
      |                            ^
      | (calls)                    | (unwinding
      v                            |  goes this
[C++ function `may_throw`]         |  way)
      |                            |
      +--- C++ function throws ----+
```

If `may_throw` does throw an exception, `b` will be dropped. Otherwise, `5`
will be printed.

## Changes since the RFC

The implementation is as specified in the RFC, but there have been a few
considerations not directly addressed in the text of the RFC.

### Additional ABI strings

The RFC, as written, only introduces the `"C-unwind"` ABI string; the others
[listed above](#summary) were introduced later. Their semantics are exactly as
one would expect: they are identical to their non-`unwind` counterparts except
when an unwind reaches their ABI boundary, in which case the behavior is the
same as the RFC specifies for `"C-unwind"`.

Similarly, the new rules for `extern "C"` regarding unwinding are now applied
to all non-`unwind` ABI strings other than `"Rust"`.

### Double unwinding

[PR #92911][pr-double-unwind] ensured that the following scenarios both safely
abort the program:

* multiple foreign (e.g. C++) exceptions
* a `panic` occurring while a foreign exception is unwinding the stack, or
  vice-versa

### Mixing panic modes

We addressed a [soundness hole][issue-mixed-panic] regarding the behavior when
a foreign exception enters the Rust runtime via a crate compiled with
`panic=unwind`, but escapes into a crate compiled with `panic=abort`. We
decided to [prohibit][pr-fix-mixed-panic] any crate from being linked with the
`panic=abort` runtime if it has both of the following characteristics:

* It contains a call to an `-unwind` foreign function or function pointer
* It was compiled with `panic=unwind`

Note: `cargo` will automatically unify all crates to use the same `panic`
runtime, so this prohibition does not apply to projects compiled with `cargo`.

[PR #97235][pr-fix-mixed-panic] implemented this prohibition.

### Unresolved questions

None of the [unresolved questions][rfc-unresolved] have been resolved.
Specifically, the following are still considered undefined behavior:

* Letting a foreign exception unwind a Rust frame that calls `catch_unwind`
* Calling `pthread_exit` or `longjmp` in such a way that Rust frames are
  deallocated

## Tests

### Codegen

There is [a folder of codegen tests][codegen-unwind] for the new ABI strings.

Additional codegen tests are:

* [panic=abort][codegen-extra-1]
* [extern-exports][codegen-extra-2]
* [extern-imports][codegen-extra-3]

### `"C"` guaranteed to abort on panic

[This test][abort-on-panic] verifies that an `extern "C"` function that
`panic!`s will always abort if the panic would otherwise "escape".

### Full examples

Full example projects mixing Rust code with non-Rust code are:

* [a `panic` escaping into C][panic-into-c]
* [a C++ exception escapig into Rust][cpp-throw-into-rust]
* [double unwinding][double-unwind]

### Open PR

[PR #97235][pr-fix-mixed-panic] introduces tests for the [mixed-panic-modes
issue](#mixing-panic-modes).

## Documentation

Here are documentation PRs for

* [the Reference][reference]
* [the Rustonomicon][nomicon]

I have not yet suggested any changes for the Book, but it may be appropriate to
add something like this:

> Usually you want the non-unwind versions of ABI strings, but if you are
> messing around with FFI and unwinding, refer to the Nomicon.

I don't think Rust by Example needs to be updated.


[rfc-text]: https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md
[rfc-table]: https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md#abi-boundaries-and-unforced-unwinding
[rfc-unresolved]: https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md#unresolved-questions
[pr-fix-mixed-panic]: https://github.com/rust-lang/rust/pull/97235/files
[pr-double-unwind]: https://github.com/rust-lang/rust/pull/92911
[issue-mixed-panic]: https://github.com/rust-lang/rust/issues/96926
[codegen-unwind]: https://github.com/rust-lang/rust/tree/master/src/test/codegen/unwind-abis
[codegen-extra-1]: https://github.com/rust-lang/rust/blob/master/src/test/codegen/unwind-and-panic-abort.rs
[codegen-extra-2]: https://github.com/rust-lang/rust/blob/master/src/test/codegen/unwind-extern-exports.rs
[codegen-extra-3]: https://github.com/rust-lang/rust/blob/master/src/test/codegen/unwind-extern-imports.rs
[panic-into-c]: https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/c-unwind-abi-catch-lib-panic
[cpp-throw-into-rust]: https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/foreign-exceptions
[double-unwind]: https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/foreign-double-unwind
[abort-on-panic]: https://github.com/rust-lang/rust/blob/master/src/test/ui/panics/abort-on-panic.rs
[nomicon]: https://github.com/rust-lang/nomicon/pull/365
[reference]: https://github.com/rust-lang/reference/pull/1226

**Comment by joshtriplett** [comment]
Shall we stabilize the `extern "C-unwind"` and other `-unwind` calling conventions? This change will leave `extern "C"` unchanged for now, but have the existing feature gate continue to opt into the new behavior on nightly. We'll do a separate change later to make `extern "C"` and similar not permit unwinding.

@rfcbt merge

**Comment by joshtriplett** [comment]
Shall we stabilize the `extern "C-unwind"` and other `-unwind` calling conventions? This change will leave `extern "C"` unchanged for now, but have the existing feature gate continue to opt into the new behavior on nightly. We'll do a separate change later to make `extern "C"` and similar not permit unwinding.

@rfcbot merge

**Comment by joshtriplett** [comment]
Also, for future reference, feature gates should not normally change behavior in any way, they should only gate access to behavior. That helps avoid adding complexity to stabilizations.

**Comment by nbdd0121** [comment]
Note that `vectorcall-unwind` and `thiscall-unwind` should not be stabilised; I'll change them to gate under `vectorcall` and `thiscall` instead.

**Comment by nikomatsakis** [comment]
@rfcbot fcp reviewed

I want to give a huge shout-out to @BatmanAoD for stewarding this work for so long. Also to @cratelyn, @Amanieu, @nbdd0121, @alexcrichton and others who participated in the implementation. It's been a slog!

**Comment by nikomatsakis** [comment]
@rfcbot concern docs

I did an initial read of the documentation and left some comments. I'm filing a concern until we have approved the docs PR.

**Comment by nikomatsakis** [comment]
@BatmanAoD also, we discussed in lang-team meeting that it'd be great to have a short description of the changes to unwinding that are being made here for release notes.

**Comment by tmandry** [comment]
I'm a slightly weirded out by the fact that `extern "C-unwind"` refers to two very different features (in my mind):

* Passing Rust exceptions through `Rust -> C++ [*] -> Rust [*]` call stacks
* Passing foreign exceptions through `C++ -> Rust [*] -> C++ [*]` call stacks

Where `A -> B` means "A calling into B". IIUC, `[*]` marks where you must have an `extern "C-unwind"` function declaration or pointer declared in Rust for that function. (Substitute C and C++ for any ABI and language pair that has exceptions.)

The RFC says a goal is to allow Rust to choose a different unwinding mechanism from the "native" one used by C/C++, but having one syntax for both feels a bit like working backwards from the fact that they're the same today.

---

That being said, I can see how in an environment where you are mixing the two languages freely, you would want functions that support _both_. Let's work out an example where we might see interleaving from both languages, and either can throw an exception:

`Rust -> C++ catch(...) [R] -> Rust [CR] -> C++ [CR] -> Rust [CR] -> C++ [C]`

At minimum you would need all the functions marked `[C]` to support C++ unwinding through them, and all the functions marked `[R]` to support Rust unwinding through them. For the vast majority of functions at language boundaries in a real-world situation, you want both (`[CR]`).

But this is actually UB, because we might pass a Rust exception through the `catch` of the second frame. (If we added a Rust `catch_unwind` farther down to prevent that, a C++ exception passing through _there_ might cause UB). So for now we must avoid this situation entirely and be clear about which types of exceptions might go through a particular function in a particular context.

If one day we think we'll support this in some way that isn't UB, it makes more sense to me to have a single syntax that supports both kinds of exceptions.

Can someone tell me if I'm conceptualizing of this properly, or am I totally confused?

**Comment by nbdd0121** [comment]
Even if Rust have a totally different unwinding mechanism, it must convert the unwinding to a way that C++ can understand. So whenever control flow leaves Rust `c-unwind` function, it needs to convert this to "native" unwind and then whenever control flow leaves a foreign `c-unwind` function, it needs to convert this from "native" unwind to Rust unwind.

i.e. `c-unwind` simply means that this function uses native unwind.

**Comment by tmandry** [comment]
@nbdd0121 Thanks, this helps! So:

* ABI (extern) corresponds to the unwind mechanism (e.g. Rust or C++)
* panic/catch_unwind, throw/catch correspond to the language that produced the exception (e.g. Rust or C++)
* Any combination of these is possible

And the above interleaving scenario is still UB, for the reasons originally stated.

**Comment by tmandry** [comment]
This has enough checks to move forward once the docs comments are addressed.

**Comment by nagisa** [comment]
> i.e. c-unwind simply means that this function uses native unwind.

Hm. What does native unwind even mean? I understand both your comment and RFC “quoted” this, and the RFC attempted to describe it in terms of `-fexceptions`… This I guess for `-gnu` targets means the `_CxxPersonality*`, the uwtables scheme this can stomach and all the surrounding related infrastructure… What would it mean for targets where the unwinding mechanism in C/C++ is not (fully) determined by any particular portion of the triple (like for the `-unknown-none` targets?) Do we disallow `-unwind` ABIs there? Or does it just happen to be “whatever LLVM does?”

---

Another point This ABI reads more like a `(lib)c-unwind` rather than “C abi + native unwind” to me. That is, lets say we add support for swift’s, Zig’s or some other language’s ABI. Does `zig-unwind` or `swift-unwind` continue making sense for the behaviour we might want to achieve? Is this still using the “native” (C++-ish) unwinding mechanism, despite the intent being to interface with `zig` and `swift` (which might be using alternate unwinding schemes)?

**Comment by BatmanAoD** [comment]
@nagisa The RFC first mentions "native" unwinding as:

>  the native exception mechanisms in GCC, LLVM, and MSVC.

So, yes, it is effectively "whatever LLVM does" when using `rustc`. I think a valid alternative way to say this is "native unwinding is whatever C++ exceptions do" on a given platform and with a given compiler-backend.

> lets say we add support for swift’s, Zig’s or some other language’s ABI. Does zig-unwind or swift-unwind continue making sense for the behaviour we might want to achieve?

I'm not sure, but also, I don't know that it matters right now, though I do think it's a good question to be cognizant of in the long term.

According to [this post](https://faultlore.com/blah/swift-abi/#esoterica), "Swift’s exceptions... are implemented more like Result and less like unwinding," which I take to mean that "native unwinding" (in the sense we're talking about here) isn't used at all.

Zig appears to use C++ unwinding (once again presumably doing "what LLVM does"), but the behavior appears to not yet be stable or fully well-defined (based on [this](https://github.com/ziglang/zig/issues/11422) and related issues).

**Comment by joshtriplett** [comment]
@nikomatsakis It looks like the [reference PR](https://github.com/rust-lang/reference/pull/1226#pullrequestreview-1229631065) has been updated; can you check it over please?

**Comment by cratelyn** [comment]
somewhat belatedly, i want to send a warm thanks to everybody that helped me out when implementing this RFC. it was my first time contributing to `rustc`, and i hope to have time to do that again soon. :heart: 

i'm excited to see this entering the FCP, and moving towards stabilization.

**Comment by BatmanAoD** [comment]
I couldn't find a link in this thread to the actual stabilization PR, so here it is: https://github.com/rust-lang/rust/pull/106075

It is currently pending review.

**Comment by RalfJung** [comment]
How long do we want to wait before completely landing this feature, and turning on the abort-on-unwind shim?

Not doing that has caused a UB bug in the standard library: https://github.com/rust-lang/rust/issues/112285.

**Comment by nbdd0121** [comment]
We need to wait for all crates that depend on ability to do FFI-unwind to upgrade to `extern "C-unwind"` before we can turn it on, and the ABI stabilisation hasn't yet hit stable yet.

**Comment by BatmanAoD** [comment]
@RalfJung @nbdd0121 We definitely need to give _some_ time for users to switch to the new ABI, but I don't think waiting for "all crates" to upgrade is reasonable. I think that ideally it would only be 1 or maybe 2 releases between `"c-unwind"` being stabilized and `"C"` behavior changing. Toward this end, I've been meaning to ping the release team and ensure they know that the 1.71 release notes should include verbiage to this effect, i.e. that the current plan is to change the `"C"` behavior very soon, possibly as early as version 1.72.

**Comment by RalfJung** [comment]
Seems prudent to wait at least 2 releases (i.e., one full stable train).

**Comment by BatmanAoD** [comment]
@RalfJung Sounds good to me! I will draft something for the 1.71 release notes accordingly (i.e. saying that we plan to change the behavior in 1.73).

**Comment by BatmanAoD** [comment]
@RalfJung @nbdd0121 @nikomatsakis @joshtriplett 

What is the next step for the `extern "C"` change? Do I need to write a separate stabilization report? Gary, would you be willing to write the PR to remove the feature flag, since you made the first stabilization PR?

Since 1.72 will be released this Thursday, it may be too late to make it in time for 1.73, but I'd like to get it in for 1.74 if possible.

**Comment by RalfJung** [comment]
Yeah a stabilization report for the unwind guards would be needed, describing what would change when the feature flag becomes "always on" and the context needed to understand that change.

**Comment by nbdd0121** [comment]
> Gary, would you be willing to write the PR to remove the feature flag, since you made the first stabilization PR?

Sure

**Comment by BatmanAoD** [comment]
# Request for Stabilization

This is a request for *full* stabilization of the `c_unwind` feature, 
[RFC-2945][rfc-text]. The behavior of non-`"Rust"`, non-`unwind` ABIs (such as
`extern "C"`) will be modified to close soundness holes caused by permitting
stack-unwinding to cross FFI boundaries that do not support unwinding.

## Summary

When using `panic=unwind`, if a Rust function marked `extern "C"` panics (and
that panic is not caught), the runtime will now abort.

Previously, the runtime would simply attempt to unwind the caller's stack, but
the behavior when doing so was undefined, because `extern "C"` functions are
optimized with the assumption that they cannot unwind (i.e. in `rustc`, they
are given the LLVM `nounwind` annotation).

This affects existing programs. If a program relies on a Rust panic "escaping"
from `extern "C"`:
* It is currently unsound.
* Once this feature is stabilized, the program will crash when this occurs,
  whereas previously it may have appeared to work as expected.
* Replacing `extern "x"` with `extern "x-unwind"` will produce the intended
  behavior without the unsoundness.

The behavior of function calls using `extern "C"` is unchanged; thus, it is
still undefined behavior to call a C++ function that throws an exception using
the `extern "C"` ABI, even when compiling with `panic=unwind`.

## Changes since the RFC

None.

### Unresolved questions

This feature does not affect the behavior of `pthread_exit` or `longjmp`, so
stabilization will neither break existing code using these features nor provide
any soundness guarantee.

## Tests

### `"C"` guaranteed to abort on panic

[This test][abort-on-panic] verifies that an `extern "C"` function that
`panic!`s will always abort if the panic would otherwise "escape".

## Documentation

The documentation PRs for this feature did not distinguish between the initial
partially-stabilized behavior and the fully-stabilized behavior, so they
already include the modifications to non-`unwind` ABIs.

* [The Rustonomicon PR][nomicon] has already been merged.
* [The Reference PR][reference] is still awaiting approval.

I've also [opened an issue][book-issue] to discuss adding a brief note about
this feature in the Book.


[rfc-text]: https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md
[abort-on-panic]: https://github.com/rust-lang/rust/blob/master/tests/ui/panics/abort-on-panic.rs
[nomicon]: https://github.com/rust-lang/nomicon/pull/365
[reference]: https://github.com/rust-lang/reference/pull/1226
[book-issue]: https://github.com/rust-lang/book/issues/3729

**Comment by BatmanAoD** [comment]
Opened a separate issue (#115285) at @joshtriplett's request, to enable re-using the FCP bot.
